### ISSUE TITLE
Deterministic dataloading setup that covers the entire data space "n" with none left behind

### ISSUE THREAD

**Issue by danielajisafe** [webdataset/webdataset#332]
Webdataset was used to train a large vision [model](https://github.com/shubham-goel/4D-Humans) on about 5 million images. Due to the large file size, I will like to simulate the same training setup and cache the model's predictions on all images. Is there a way to get a deterministic setup that would cover all 5 million images (during dataloading) with none left behind? I have tried turning off/commenting out,

_resampled, 
dataset.shuffle, 
shardshuffle,
Pytorch dataloader's shuffle, 
set different random Seeds, 
n_workers=1,
set prob to None in wds.RandomMix, 
export PYTHONHASHSEED=0_

But none of the above works. @tmbdev Your timely help would be appreciated.

**Comment by tmbdev** [comment]
It's unclear to me what you are asking. If you use this:

```
dataset = WebDataset("data-{000000..000999}.tar")
for sample in dataset:
	...
```

then the iterator will deterministically iterate through each sample in sequence, from beginning to end, each and every time.  There is nothing complicated or nondeterministic happening behind the scenes: WebDataset is just a convenience notation for a series of simple iterators. However, the PyTorch Dataloader is generally non-deterministic.

It sounds like what you may want to do is large scale parallel inference. Your best bet for that is to parallelize over shards. You can use whatever parallel library you like; using Ray, this looks something like:

```
def process_shard(input_shard):
	output_shard = ... compute output shard name...
    src = wds.WebDataset(input_shard).decode("PIL")
    snk = wds.TarWriter(output_shard)
    for sample in src:
      sample["cls"] = classifier(sample["jpg"])
      ... do more/other stuff, maybe delete the image...
      snk.write(sample)
   src.close();
   snk.close()
   return output_shard

shards = list(braceexpand("data-{000000..000999}.tar"))
results = ray.map(process_shard, shards)
ray.get(results)
```

Writing a separate parallel inference command and parallelizing over shards is generally faster than building on top of the DataLoader pipeline you use during training.

Note that you do not have to write the classification results as WebDataset files; you can also just write them as JSON files or send them off to a database.
